src/RamRelation.h

```diff
@@ -314,6 +322,86 @@ class RamRelation {
         num_tuples++;
     }
 
+    /** insert a new tuple to table, possibly more than one tuple depending on relation type */
+    void insert(const RamDomain* tuple) {
+        // TODO: (pnappa) an eqrel check here is all that appears to be needed for implicit additions
+        if (id.isEqRel()) {
+            // TODO: future optimisation would require this as a member datatype
+            // brave soul required to pass this quest
+            // // specialisation for eqrel defs
+            // std::unique_ptr<binaryrelation> eqreltuples;
+            // in addition, it requires insert functions to insert into that, and functions
+            // which allow reading of stored values must be changed to accommodate.
+            // e.g. insert =>  eqRelTuples->insert(tuple[0], tuple[1]);
+
+            // for now, we just have a naive & extremely slow version, otherwise known as a O(n^2) insertion
+            // ):
+
+            // store all values that will be implicitly relevant to the two that we will insert
+            std::list<const RamDomain*> relevantStored;
```

Comment:
Vectors are O(1) amortised for appending, so overall the complexity is the same. The advantages are smaller constant factors for memory and speed, and potential benefits from contiguous memory. I'm not sure how a BlockList compares, but it looks a bit like std::deque so I expect it has similar advantages, with a little more constant overhead than vector and big advantages for other operations.